src/conversions.jl

```diff
@@ -1,12 +1,88 @@
-Base.promote_rule(::Type{T1}, ::Type{T2}) where {T1<:AbstractGray,T2<:AbstractGray} = Gray{promote_type(eltype(T1), eltype(T2))}
-Base.promote_rule(::Type{T1}, ::Type{T2}) where {T1<:AbstractRGB,T2<:AbstractRGB} = RGB{promote_type(eltype(T1), eltype(T2))}
 
-Base.promote_rule(::Type{C3}, ::Type{Cgray}) where {C3<:Color3,Cgray<:AbstractGray} = base_colorant_type(C3){promote_type(eltype(C3), eltype(Cgray))}
-Base.promote_rule(::Type{C3}, ::Type{Cagray}) where {C3<:Color3,Cagray<:AbstractAGray} = alphacolor(base_colorant_type(C3)){promote_type(eltype(C3), eltype(Cagray))}
-Base.promote_rule(::Type{C3}, ::Type{Cgraya}) where {C3<:Color3,Cgraya<:AbstractGrayA} = coloralpha(base_colorant_type(C3)){promote_type(eltype(C3), eltype(Cgraya))}
+function _promote_et(::Type{C1}, ::Type{C2}) where {C1<:Colorant, C2<:Colorant}
+    T1 = isconcretetype(eltype(C1)) ? eltype(C1) : eltypes_supported(C1)
+    T2 = isconcretetype(eltype(C2)) ? eltype(C2) : eltypes_supported(C2)
+    promote_type(T1, T2)
+end
+
+_with_et(C::UnionAll, et) = isconcretetype(et) ? C{et} : C
+function _transparentcolor_with_et(::Type{C0}, et) where {N, C0<:TransparentColorN{N}}
+    if isconcretetype(et)
+        C0 <: AlphaColor && return AlphaColor{C,et,N} where {C<:Color{et,N-1}}
+        C0 <: ColorAlpha && return ColorAlpha{C,et,N} where {C<:Color{et,N-1}}
+        return TransparentColor{C,et,N} where {C<:Color{et,N-1}}
+    else # remove the constraint of eltype `T` of base color type `C`
+        C0 <: AlphaColor && return AlphaColor{C,T,N} where {C<:ColorN{N-1}, T}
```

Comment:
Should we use
```julia
AlphaColor{C,T,N} where {T,C<:Color{T,N-1}}
```
Optional; since `C` is unspecified I doubt it matters a lot.

Reply:
Please excuse my lack of explanation. It looks like a difference of point of view between you and me. (cf. https://github.com/JuliaGraphics/ColorTypes.jl/pull/161#discussion_r379234848)

In fact, the eltype `T` of `TransparentColor` and the eltype of base type `C` are chained (i.e. the same). However, the definitions of `TransparentColor`, `AlphaColor`, and `ColorAlpha` have no such constraints.

This difference is obvious when using "plain" types. However, the extra constraints can be confusing when using the typealiases.
```julia
julia> base_colorant_type(Transparent3) == Transparent3
false
```

We might want to change the type definitions (i.e. introduce the constraints), but I can't imagine the impact.

Reply:
Yes, back when ColorTypes was written it wasn't possible to chain them. Now it is, but there's still one thing I want that isn't: chaining the `N` between the Color and the TransparentColor. If it does become possible (https://github.com/JuliaLang/julia/issues/8322), I'd say we should change the type definitions, but I am not sure it's worth it now just to go through it all over again.

Reply:
For consistency with `base_colorant_type`, I added the element type constraint.
https://github.com/JuliaGraphics/ColorTypes.jl/blob/3dfb0045e49d78838d4f530003e43f47484cd630/src/conversions.jl#L15-L22